crates/span/src/ast_id.rs

```diff
@@ -152,126 +548,345 @@ impl AstIdMap {
+                                    Some(node),
+                                    "left a BlockExpr we never entered"
+                                );
+                            }
+                        }
+                    }
+                }
+            });
+            std::mem::swap(&mut curr_layer, &mut next_layer);
+            assert!(blocks.is_empty(), "didn't leave all BlockExprs");
+        }
+
+        res.ptr_map = hashbrown::HashTable::with_capacity(res.arena.len());
+        res.id_map = hashbrown::HashTable::with_capacity(res.arena.len());
+        for (idx, (ptr, ast_id)) in res.arena.iter() {
+            let ptr_hash = hash_ptr(ptr);
+            let ast_id_hash = hash_ast_id(ast_id);
+            match res.ptr_map.entry(
+                ptr_hash,
+                |idx2| *idx2 == idx,
+                |&idx| hash_ptr(&res.arena[idx].0),
+            ) {
+                hashbrown::hash_table::Entry::Occupied(_) => unreachable!(),
+                hashbrown::hash_table::Entry::Vacant(entry) => {
+                    entry.insert(idx);
+                }
             }
-        });
-        res.map = hashbrown::HashTable::with_capacity(res.arena.len());
-        for (idx, ptr) in res.arena.iter() {
-            let hash = hash_ptr(ptr);
-            match res.map.entry(hash, |&idx2| idx2 == idx, |&idx| hash_ptr(&res.arena[idx])) {
+            match res.id_map.entry(
+                ast_id_hash,
+                |idx2| *idx2 == idx,
+                |&idx| hash_ast_id(&res.arena[idx].1),
+            ) {
                 hashbrown::hash_table::Entry::Occupied(_) => unreachable!(),
                 hashbrown::hash_table::Entry::Vacant(entry) => {
                     entry.insert(idx);
                 }
             }
         }
         res.arena.shrink_to_fit();
-        res
+        return res;
+
+        fn parent_of(parent_idx: Option<ArenaId>, res: &AstIdMap) -> Option<&ErasedFileAstId> {
+            let mut parent = parent_idx.map(|parent_idx| &res.arena[parent_idx].1);
+            if parent.is_some_and(|parent| parent.kind() == ErasedFileAstIdKind::ExternBlock as u32)
+            {
+                // See the comment on `ErasedAssocItemFileAstId` for why is this.
+                // FIXME: Technically there could be an extern block inside another item, e.g.:
+                // ```
+                // fn foo() {
+                //     extern "C" {
+                //         fn bar();
+                //     }
+                // }
+                // ```
+                // Here we want to make `foo()` the parent of `bar()`, but we make it `None`.
+                // Shouldn't be a big deal though.
+                parent = None;
+            }
+            parent
+        }
     }
 
     /// The [`AstId`] of the root node
     pub fn root(&self) -> SyntaxNodePtr {
-        self.arena[Idx::from_raw(RawIdx::from_u32(0))]
+        // FIXME: Seems redundant to store the root.
```

Comment:
Reason for storing it was simpler code overall (no special casing needed for the root)